### PR TITLE
Decoupled projects: Contract changes

### PR DESCRIPTION
**Pull request by topocount**
Addresses #1875 and #1897

This PR adds functions to the projects contract for:
- creating decoupled projects that can include an IPFS hash
- creating and modifiying issues that don't yet contain bounties

This PR also contains
- gas optimizations to allow these issues to fit in the contract
- the removal of the standard bounties bytecode check to make room for these features, since `isContract` should be sufficient

**Note**
- [x] merge #1908 before this PR

### CODE REVIEW

File: apps/projects/test/Projects.test.js [hunk 1]
```diff
@@ -1342,7 +1443,7 @@ contract('Projects App', accounts => {
         })
       })
 
-      it('cannot update bounties contract with contract of invalid size', async () => {
```
Comment (Quazia) [inline apps/projects/test/Projects.test.js:1345]: Might as well just remove this text I think?

File: apps/projects/contracts/Projects.sol [hunk 2]
```diff
@@ -321,10 +298,14 @@ contract Projects is AragonApp, DepositableStorage {
      * @param _repoId The id of the repo in the projects registry
      * @return index the repo registry index
```
Comment (Quazia) [inline apps/projects/contracts/Projects.sol]: Return is wrong now

**Review by Quazia — DISMISSED** [review]
Looks good, in integration we'll need to check for all instances of `IssueUpdated` and the parsing of return data from `getIssue` I'm not seeing anything else that should need changing.

**Comment by rkzel** [comment]
Specification suggests that there should be a fuction to edit project as well as add/remove. Editing will involve only replacing ipfs hash saved in`_projectData`, so all we need is a simple function like this:

```
    function updateRepo(bytes32 _repoId, string _repoData) external auth(....?)
     {
        require(isRepoAdded(_repoId), ERROR_REPO_MISSING);
        uint rowToUpdate = repos[_repoId].index;
        repoIndex[repoIndexLength] = _repoId;
        repos[rowToUpdate].repoData = _repoData;
        emit Repo....?
    }
```
Probably a new right and event to emit  need to be added as well - what do you think, @topocount ? Alternatively if it's too complicated for current timeline, maybe we leave it for now and not allow editing projects at all?

File: apps/projects/contracts/Projects.sol [hunk 3]
```diff
@@ -319,12 +297,16 @@ contract Projects is AragonApp, DepositableStorage {
     /**
      * @notice Get an entry from the registry.
      * @param _repoId The id of the repo in the projects registry
-     * @return index the repo registry index
+     * @return index the repo registry index, number of open bounties, decoupled bool, repo metadata
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:300]: It is preferable to document `@return` params into separate lines each one

File: apps/projects/contracts/Projects.sol [hunk 4]
```diff
@@ -357,20 +339,25 @@ contract Projects is AragonApp, DepositableStorage {
 // Repository functions
 ///////////////////////
     /**
-     * @notice Add repository
+     * @notice Add new Project: `_projectData`.
      * @param _repoId The id of the repo in the projects registry
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:343]: We should document the rest of params as well

File: apps/projects/contracts/Projects.sol [hunk 5]
```diff
@@ -578,6 +567,25 @@ contract Projects is AragonApp, DepositableStorage {
...
+     * @param _data the provided proof of fulfillment
+     */
+    function fulfillBounty(
+        address _sender,
+        uint _bountyId,
+        address[] _fulfillers,
+        string _data
+    ) external
+    {
+        revert();
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:586]: If we just want to display the radspec, why revert? Also this should be a pure function

Comment (ottodevs) [reply apps/projects/contracts/Projects.sol:586]: Revert will cause non 0 exit status and that maybe non gracefully handled by the frontend

File: apps/projects/contracts/Projects.sol [hunk 6]
```diff
@@ -781,40 +754,68 @@ contract Projects is AragonApp, DepositableStorage {
...
+    function _setRepo(
+        bytes32 _repoId,
+        bool _decoupled,
+        string _repoData
+    ) internal
+    {
+        repos[_repoId].repoData = _repoData;
+        if (!isRepoAdded(_repoId)) {
+            repoIndex[repoIndexLength] = _repoId;
+            repos[_repoId].index = repoIndexLength++;
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:766]: Are we accounting on a potential overflow here?

File: apps/projects/contracts/Projects.sol [hunk 7]
```diff
@@ -781,40 +754,68 @@ contract Projects is AragonApp, DepositableStorage {
...
+        uint256[] _deadlines,
+        uint256[] _tokenTypes,
+        address[] _tokenContracts,
+        string _ipfsAddresses,
+        string _description,
+        bool _open
+    ) internal
+    {
+        // ensure the transvalue passed equals transaction value
+        //checkTransValueEqualsMessageValue(msg.value, _bountySizes,_tokenBounties);
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:787]: Is this comment relevant anymore or just dead code? We should remove to avoid confusion

File: apps/projects/contracts/Projects.sol [hunk 8]
```diff
@@ -781,40 +754,68 @@ contract Projects is AragonApp, DepositableStorage {
...
+        string memory ipfsHash;
+        uint standardBountyId;
+        require(bytes(_ipfsAddresses).length == (CID_LENGTH * _bountySizes.length), ERROR_CID_LENGTH);
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:790]: We should test how it behaves with `_bountySizes.length == 0` as it could bypass the condition in edge cases

File: apps/projects/contracts/Projects.sol [hunk 9]
```diff
@@ -926,31 +927,48 @@ contract Projects is AragonApp, DepositableStorage {
         uint256 _bountySize,
         string _ipfsHash
     ) internal
+    {
+        //Issue storage issue = repos[_repoId].issues[_issueNumber];
+        //require(isRepoAdded(_repoId), ERROR_REPO_MISSING);
```
Comment (ottodevs) [inline apps/projects/contracts/Projects.sol:932]: Is this comment relevant? we should clean in case not

File: apps/projects/contracts/test/BountiesEvents.sol [hunk 10]
```diff
@@ -3,32 +3,38 @@ pragma solidity ^0.4.24;
...
+     * Events
+     */
+
+    event BountyIssued(uint _bountyId, address _creator, address [] _issuers, address [] _approvers, string _data, uint _deadline, address _token, uint _tokenVersion);
+    event ContributionAdded(uint _bountyId, uint _contributionId, address  _contributor, uint _amount);
+    event ContributionRefunded(uint _bountyId, uint _contributionId);
+    event ContributionsRefunded(uint _bountyId, address _issuer, uint [] _contributionIds);
+    event BountyDrained(uint _bountyId, address _issuer, uint [] _amounts);
+    event ActionPerformed(uint _bountyId, address _fulfiller, string _data);
+    event BountyFulfilled(uint _bountyId, uint _fulfillmentId, address  [] _fulfillers, string _data, address _submitter);
```
Comment (ottodevs) [inline apps/projects/contracts/test/BountiesEvents.sol:32]: Some spacing between array brackets and types seem wrong with inconsistently 1 or 2 spaces, maybe we should try to be consistent about it, I vote for no spacing between type and brackets
```suggestion
    event BountyFulfilled(uint _bountyId, uint _fulfillmentId, address[] _fulfillers, string _data, address _submitter);
```

Some for the rest of lines on this block

File: apps/projects/test/Projects.test.js [hunk 11]
```diff
@@ -1,4 +1,4 @@
-/* global artifacts, assert, before, beforeEach, contract, context, expect, it, web3 */
+/* global artifacts, assert, before, beforeEach, contract, context, expect, it, web3, xit */
```
Comment (ottodevs) [inline apps/projects/test/Projects.test.js:1]: `xit` should actually not need to be defined here, as tests should only be skipped during deployment

File: apps/projects/test/Projects.test.js [hunk 12]
```diff
@@ -872,6 +848,46 @@ contract('Projects App', accounts => {
...
+            [9],
+            { from: bountyManager }
+          )
+
+          await app.reviewSubmission(
+            repoId,
+            1,
+            1,
+            true,
+            'QmbUSy8HCn8J4TMDRRdxCbK2uCCtkQyZtY6XYv3y7kLgDl',
```
Comment (ottodevs) [inline apps/projects/test/Projects.test.js:873]: I would move most of these arbitrary magic values to constante to have a more manageable/less error prone test suite, but that's only a personal preference

File: apps/projects/test/Projects.test.js [hunk 13]
```diff
@@ -872,6 +848,46 @@ contract('Projects App', accounts => {
...
+          )
+          
+          const bounty1 = await app.getIssue(repoId, 1)
+          assert.deepEqual(
+            bounty1,
+            [ 
+              new web3.BigNumber(56),
+              new web3.BigNumber(15),
+              new web3.BigNumber(0),
+              '0xffffffffffffffffffffffffffffffffffffffff',
```
Comment (ottodevs) [inline apps/projects/test/Projects.test.js:885]: We should use ANY_ADDR (line 16) constant in places like here, otherwise, we define it at all?

File: apps/projects/test/Projects.test.js [hunk 14]
```diff
@@ -1156,6 +1170,122 @@ contract('Projects App', accounts => {
       })
     })
 
+    context('decoupled project', () => {
+      let repoId
+      beforeEach(async () => {
+        await app.setRepo('1', true, 'abc')
```
Comment (ottodevs) [inline apps/projects/test/Projects.test.js:1176]: We should use real repo ids as in other test cases to avoid false negatives when repoId length does not find in the contract field

**Review by ottodevs — CHANGES_REQUESTED** [review]
Changed base branch to merge, to properly see the exact changes going on here

Most of the contract and tests look fine for me, and they are passing. I am a bit concerned about opening potential security vectors, as we are not auditing this new code on this review.

Just some minor details to comment about:

It would be preferable to properly document each function that was added or changed
I would just take advantage of this work to keep the contract clean and audit-ready by properly documenting this new code.

I would also bet to improve the tests suite and make it more readable/ less error prone by making use of constants and avoiding magic numbers and text everywhere.

**Comment by topocount** [comment]
> Changed base branch to merge, to properly see the exact changes going on here

We're working out of the https://github.com/AutarkLabs/open-enterprise/tree/decentralized-project-integration banch, and this branch has already been merged in there. I'll make your requested changes in there and redirect https://github.com/AutarkLabs/open-enterprise/tree/issue-querying to that branch as well.

File: apps/projects/contracts/Projects.sol (hunk 5 above)
Comment (topocount) [reply apps/projects/contracts/Projects.sol:586]: This function should never execute. It's just a workaround for getting the radspec showing up in the frontend. BY forcing it to revert, it effectively doesn't exist and is an invalid call on the contract

File: apps/projects/contracts/Projects.sol (hunk 8 above)
Comment (topocount) [reply apps/projects/contracts/Projects.sol:790]: if someone passes in _bountySizes with a length of zero, they' just be wasting gas, since the for loop won't be entered

File: apps/projects/test/Projects.test.js (hunk 14 above)
Comment (topocount) [reply apps/projects/test/Projects.test.js:1176]: Projects will be using integers for decoupled repoIds going forward. They'll still get hashed by the frontend before passing them in, but that's really just a messier magic number in the context of tests

**Comment by topocount** [comment]
This branch has been merged into PR #1941. The review above has been addressed there
